python/ml4ir/base/pipeline.py

```diff
@@ -265,9 +269,209 @@ def get_relevance_model(self, feature_layer_keys_to_fns={}) -> RelevanceModel:
+        """
+        Aggregate results of the k-fold runs and perform t-test on the results.
+        """
+        raise NotImplementedError
+
+    def create_folds(self, num_folds, fold_id, include_testset_in_kfold, all_data):
+        """
+        Create training, validation and test set according to the passed fold number.
+
+        Parameters
+        ----------
+        num_folds : int
+            Total number of folds
+        fold_id : int
+            current fold number
+        include_testset_in_kfold: bool
+            whether to include testset in creating folds
+        all_data: Tensorflow Dataset
+            the dataset used to create folds
+
+        Returns
+        -------
+        training: Tensorflow Shard Dataset
+            training set corresponding to the current fold number
+        validation: Tensorflow Shard Dataset
+            validation set corresponding to the current fold number
+        validation: Tensorflow Shard Dataset
+            test set corresponding to the current fold number
+        """
+        test = None
+        training_idx = list(range(num_folds))
+        if include_testset_in_kfold:
+            validation = all_data.shard(num_folds, fold_id)
+            test_idx = fold_id + 1
+            if fold_id + 1 >= num_folds:
+                test_idx = 0
+            test = all_data.shard(num_folds, test_idx)
+            training_idx.remove(test_idx)
+        else:
+            validation = all_data.shard(num_folds, fold_id)
+        training_idx.remove(fold_id)
+        training = None
+        for j in training_idx:
+            if not training:
+                training = all_data.shard(num_folds, j)
+            else:
+                training = training.concatenate(all_data.shard(num_folds, j))
+        return training, validation, test
+
+    def merge_datasets(self, relevance_dataset, include_testset_in_kfold):
+        """
+        Concat the datasets (training, validation, test) together
+
+        Parameters
+        ----------
+        relevance_dataset : RelevanceDataset
+            RelevanceDataset object holding the training, validation and test datasets.
+        include_testset_in_kfold : bool
+            whether to include testset in creating folds
+
+
+        Returns
+        -------
+        all_data: Tensorflow Dataset
+            The final concatenated dataset.
+        """
+
+        if include_testset_in_kfold:
+            all_data = relevance_dataset.train.concatenate(relevance_dataset.validation).concatenate(
+                relevance_dataset.test)
+        else:
+            all_data = relevance_dataset.train.concatenate(relevance_dataset.validation)
+
+        # un-batch and shuffle all queries
+        all_data = all_data.unbatch()
+        # shuffling before using the shard method gives unexpected results. Should be avoided
+        # all_data = all_data.shuffle(batch_size * 2)
+        return all_data
```

Comment:
@mohazahran I think you should move these two methods to `RelevanceDataset`. Maybe even create a `KFoldDataset` class and have methods to fetch, perhaps?